### ISSUE TITLE
find a way to allow use of shellcheck with jinja2 shell script templates (ansible)

### ISSUE THREAD

**Issue by ssbarnea** [koalaman/shellcheck#1306]
More and more shell scripts are migrated from pure bash to being jinja2 templates as they are
easily process by ansible.

Sadly at this moment shellcheck cannot be used with shellscripts are jinja templates and it would be really useful to find a way to run it jinja2 templates, even if it may nit cover all jinja2 use-cases.

**Comment by ngzhian** [comment]
Any way to preprocess your jinja2 templates to feed it to ShellCheck? That way we don't have to add special cases ShellCheck for new file types that are not really shell.

**Comment by ssbarnea** [comment]
@willthames maybe you have an idea about that? Ability to lint jinja2 templates and what they would produce could be something very handy for ansible.

Clearly we would need a way to process jinja2 during runtime in order to obtain a likely result which should be passed to an external linter. One problem is to identify which linter should be called, as jinja2 can expand to almost anything.

**Comment by lonix1** [comment]
Did you find a workaround?

These days ansible is used heavily for automation, so scripts contain `{{foo}}`.

(And this triggers [SC1083](https://github.com/koalaman/shellcheck/wiki/SC1083))

**Comment by TomMD** [comment]
Saltstack is often paired with Jinja too.  It seems we could add a flag `--jinja` that will allow relaxed syntax parsing, perhaps including a new AST element for conditional portions of scripts (lifting the compile time condition into what is abstractly a run-time condition).

**Comment by micheelengronne** [comment]
Indeed, my shell scripts include a lot of jinja2 statements (`{% %}`) for template inheritance. Shellcheck is totally lost with that. I canno't pretemplate them as it breaks the flow of the CI (linters first before doing any transformation to avoid breaking transformations). So, I sadly gave up on shellchecking jinja2 files.

**Comment by lonix1** [comment]
The fix is to have an disable-then-enable directive as I proposed in #2512

**Comment by schewara** [comment]
Just stumbling into this problem myself and with some spelunking around, I discovered

- [jinjanator](https://github.com/kpfleming/jinjanator)

which looks like a promising tool to integrate or hook into shellcheck.

As Jinja is just one of many [template engines](https://en.wikipedia.org/wiki/Comparison_of_web_template_engines) it might be overkill for this project to support most of them, therefore having an option to pass or pipe the rendered output to shellcheck would be super helpful and would probably scale better.

being able to use something like

```
$ jinjanate test.sh.j2 testdata.yml | shellcheck
```

would be super helpful, but currently one would have to jump through some hoops to accomplish the same, like in the following example.

```
$ jinjanate test.sh.j2 testdata.yml > /tmp/shellcheck-test.sh
jinjanate 24.3.0, Jinja2 3.1.4

$ shellcheck /tmp/shellcheck-test.sh

In /tmp/shellcheck-test.sh line 3:
moo = "a cow"
    ^-- SC2283 (error): Remove spaces around = to assign (or use [ ] to compare, or quote '=' if literal).

For more information:
  https://www.shellcheck.net/wiki/SC2283 -- Remove spaces around = to assign ...

$ rm -f /tmp/shellcheck-test.sh
```

#2512 in addition also looks like a good alternative approach, which reminds me of Jinjas own [*raw* blocks](https://jinja.palletsprojects.com/en/3.0.x/templates/#escaping) approach.

**Comment by prahal** [comment]
Note that #2512 would not help, as "disable" does not prevent the parser from stopping the parsing. "Silencing parser errors is purely cosmetic; Any parser error found will still stop shellcheck at the point of the error."  https://github.com/koalaman/shellcheck/wiki/Directive#disable
